src/lib/OS/SyscallHandler.cc

```diff
@@ -1,15 +1,712 @@
+      break;
+    }
+    case 48: {  // faccessat
+      int64_t dfd = info.registerArguments[0].get<int64_t>();
+      uint64_t filenamePtr = info.registerArguments[1].get<uint64_t>();
+      int64_t mode = info.registerArguments[2].get<int64_t>();
+      // flag component not used, although function definition includes it
+      int64_t flag = 0;
+
+      char* filename = new char[PATH_MAX_LEN];
+      return readStringThen(
+          filename, vAddrTranslation_(filenamePtr, info.processId),
+          PATH_MAX_LEN, [=](auto length) {
+            // Invoke the kernel
+            int64_t retval = faccessat(dfd, filename, mode, flag);
+            ProcessStateChange stateChange = {
+                ChangeType::REPLACEMENT, {info.ret}, {retval}};
+            delete[] filename;
+            concludeSyscall(stateChange);
+          });
+      break;
+    }
+    case 56: {  // openat
+      int64_t dirfd = info.registerArguments[0].get<int64_t>();
+      uint64_t pathnamePtr = info.registerArguments[1].get<uint64_t>();
+      int64_t flags = info.registerArguments[2].get<int64_t>();
+      uint16_t mode = info.registerArguments[3].get<uint16_t>();
+
+      char* pathname = new char[PATH_MAX_LEN];
+      return readStringThen(
+          pathname, vAddrTranslation_(pathnamePtr, info.processId),
+          PATH_MAX_LEN, [=](auto length) {
+            // Invoke the kernel
+            uint64_t retval = openat(dirfd, pathname, flags, mode);
+            ProcessStateChange stateChange = {
+                ChangeType::REPLACEMENT, {info.ret}, {retval}};
+            delete[] pathname;
+            concludeSyscall(stateChange);
+          });
+      break;
+    }
+    case 57: {  // close
+      int64_t fd = info.registerArguments[0].get<int64_t>();
+      stateChange = {ChangeType::REPLACEMENT, {info.ret}, {close(fd)}};
+      break;
+    }
+    case 61: {  // getdents64
+      int64_t fd = info.registerArguments[0].get<int64_t>();
+      uint64_t bufPtr = info.registerArguments[1].get<uint64_t>();
+      uint64_t count = info.registerArguments[2].get<uint64_t>();
+
+      return readBufferThen(
+          vAddrTranslation_(bufPtr, info.processId), count, [=]() {
+            int64_t totalRead = getdents64(fd, dataBuffer_.data(), count);
+            ProcessStateChange stateChange = {
+                ChangeType::REPLACEMENT, {info.ret}, {totalRead}};
+            // Check for failure
+            if (totalRead < 0) {
+              return concludeSyscall(stateChange);
+            }
+
+            int64_t bytesRemaining = totalRead;
+            // Get pointer and size of the buffer
+            uint64_t iDst = bufPtr;
+            uint64_t iLength = bytesRemaining;
+            if (iLength > bytesRemaining) {
+              iLength = bytesRemaining;
+            }
+            bytesRemaining -= iLength;
+            // Write data for this buffer in 128-byte chunks
+            auto iSrc = reinterpret_cast<const char*>(dataBuffer_.data());
```

Comment:
If `dataBuffer` was changed to a vector<char> then this cast would not be needed

Reply:
Again, this applies throughout this file

Reply:
removed cast